plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -311,89 +396,192 @@ func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
+}
+
+func (c *CloudWatch) aggregateMetrics(
+	acc telegraf.Accumulator,
+	metricDataResults []*cloudwatch.MetricDataResult,
+) error {
+	namespace := formatMeasurement(c.Namespace)
+
+	for _, query := range c.queries {
 		tags := map[string]string{
 			"region": c.Region,
-			"unit":   snakeCase(*point.Unit),
 		}
 
-		for _, d := range metric.Dimensions {
-			tags[snakeCase(*d.Name)] = *d.Value
+		type metric struct {
+			fields map[string]interface{}
+			tags   map[string]string
 		}
+		thing := map[time.Time]metric{}
+		for _, result := range metricDataResults {
+			if nameMatch(query.id, *result.Id) {
+				for _, dimension := range query.metric.Dimensions {
+					tags[snakeCase(*dimension.Name)] = *dimension.Value
+				}
+				if len(result.Timestamps) != len(result.Values) {
+					log.Println("[inputs.cloudwatch] W! MISMATCHED TIMESTAMP/VALUE LENGTH!!", len(result.Timestamps), len(result.Values))
+					continue
+				}
 
-		// record field for each statistic
-		fields := map[string]interface{}{}
-
-		if point.Average != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticAverage)] = *point.Average
-		}
-		if point.Maximum != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticMaximum)] = *point.Maximum
-		}
-		if point.Minimum != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticMinimum)] = *point.Minimum
-		}
-		if point.SampleCount != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticSampleCount)] = *point.SampleCount
-		}
-		if point.Sum != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticSum)] = *point.Sum
+				for j := range result.Values {
+					if _, ok := thing[*result.Timestamps[j]]; !ok {
+						thing[*result.Timestamps[j]] = metric{fields: map[string]interface{}{*result.Label: *result.Values[j]}, tags: tags}
+					}
+					thing[*result.Timestamps[j]].fields[*result.Label] = *result.Values[j]
+				}
+			}
 		}
 
-		acc.AddFields(formatMeasurement(c.Namespace), fields, tags, *point.Timestamp)
+		for t, m := range thing {
+			acc.AddFields(namespace, m.fields, m.tags, t)
+		}
 	}
 
 	return nil
 }
 
+func nameMatch(name, id string) bool {
+	if strings.TrimPrefix(id, "average_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "maximum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "minimum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "sum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "sample_count_") == name {
+		return true
+	}
+	return false
+}
+
 /*
  * Formatting helpers
  */
-func formatField(metricName string, statistic string) string {
-	return fmt.Sprintf("%s_%s", snakeCase(metricName), snakeCase(statistic))
-}
-
 func formatMeasurement(namespace string) string {
 	namespace = strings.Replace(namespace, "/", "_", -1)
 	namespace = snakeCase(namespace)
-	return fmt.Sprintf("cloudwatch_%s", namespace)
+	return "cloudwatch_" + namespace
 }
 
 func snakeCase(s string) string {
```

Comment:
What do you think of calling this function something like `sanitizeName` or `encode` or something ask it is doing more than just `snakeCase`?

Reply:
Realizing this is terrible, but I generally don't change pre-existing working code. Maybe my justification was that it would be easier for someone to come back to and not be surprised we aren't "formatting the measurement" anymore. I'm concluding now that I'm in the camp of "improve-all-the-code".